### PR TITLE
JAMES-4085 BlobStore: introduce APIs for multi-tenancy support

### PR DESCRIPTION
**Pull request by quantranhong1999**

### CODE REVIEW

**Review by vttranlina — COMMENTED** [review]
read it

**Review by hungphan227 — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

File: server/blob/blob-api/src/main/java/org/apache/james/blob/api/BlobStore.java [hunk 1]
```diff
@@ -43,45 +45,118 @@ interface BlobIdProvider {
 
     Publisher<BlobId> save(BucketName bucketName, byte[] data, StoragePolicy storagePolicy);
 
+    default Publisher<BlobId> save(Bucket bucket, byte[] data, StoragePolicy storagePolicy) {
+        return save(bucket.bucketName(), data, storagePolicy);
+    }
```
Comment (chibenwa) [inline server/blob/blob-api/src/main/java/org/apache/james/blob/api/BlobStore.java:50]: This makes it harder to use the class: we can not double the number of methods every time we add some parameters.

I'd be happy we make it easy to get a Bucket from a BucketName and sollely depend on bucket

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by jeantil** [comment]
can you wait a bit before merging ? I will try to do a thorough review tonight. I am concerned about `getDefaultBucket` and how it is used. 
I feel that this method is dangerous as it will never account for a tenant making it easy to ignore tenancy at the callsite. if we want the blobstore and james to be multitenant, the tenancy information should probably propagate throughout all the call sites until we the point where we can determine the information.

**Comment by chibenwa** [comment]
Hi @jeantil 

> can you wait a bit before merging ?

of course ^^

> I am concerned about getDefaultBucket and how it is used.
I feel that this method is dangerous as it will never account for a tenant making it easy to ignore tenancy at the callsite. if we want the blobstore and james to be multitenant, the tenancy information should probably propagate throughout all the call sites until we the point where we can determine the information

Legitimate concerns indeed yet : 

 - the only component being currently multi-tenant is the mailbox
 - and the only multi-tenant mailbox  implementation is the PGSQL one.

The proposed changeset will then only make sense on top of that mailbox

> I am concerned about getDefaultBucket and how it is used.

We could easily say we kill that method, and make the notion of "default bucket" a caller concern

(test constant, etc...)

**Comment by chibenwa** [comment]
@quantranhong1999 can you have a look at what it takes to remove `BloBstore::getDefaultBucket`

File: server/blob/blob-api/src/main/java/org/apache/james/blob/api/BlobStoreDAO.java [hunk 2]
```diff
@@ -128,9 +129,14 @@ default Publisher<Void> save(BucketName bucketName, BlobId blobId, String data)
      * @return a successful Publisher if the bucket is deleted or did not exist
      *  otherwise an IOObjectStoreException in its error channel
      */
-    Publisher<Void> deleteBucket(BucketName bucketName);
+    Publisher<Void> deleteBucket(Bucket bucket);
 
     Publisher<BucketName> listBuckets();
```
Comment (jeantil) [inline server/blob/blob-api/src/main/java/org/apache/james/blob/api/BlobStoreDAO.java:134]: this should return Buckets to be consistent wit the changed apis

**Comment by jeantil** [comment]
I realized that this PR is the implementation of the mail that proposes the change on multi-tenancy that was sent yesterday. 
Looking at the code, jira and mail, I realized that my concerns are more at an architectural level than related to this implementation ( though the challenge to the architectural choices may render this changeset obsolete)  I thus answered with my design questions on the ML instead

**Comment by chibenwa** [comment]
Looks like the email did not went through

image

**Comment by quantranhong1999** [comment]
I mark this PR as a draft as the design for this topic likely is complicated and deserves more discussion among the community.

**Pull request closed without merging**